urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/driver/INDIProperty.java

```diff
@@ -0,0 +1,615 @@
+   *
+   * @param message An message to be sent to the client when setting the values
+   * of the property.
+   * @return The XML code to set the values of the property.
+   */
+  protected String getXMLPropertySet(String message) {
+    if (saveable) {
+      try {
+        saveToFile();
+      } catch (IOException e) {
+      }
+    }
+
+    String xml;
+
+    if (message == null) {
+      xml = getXMLPropertySetInit();
+    } else {
+      xml = getXMLPropertySetInit(message);
+    }
+
+    List<INDIElement> elem = getElementsAsList();
+
+    for (int i = 0 ; i < elem.size() ; i++) {
+      xml += elem.get(i).getXMLOneElement();
+    }
+
+    xml += getXMLPropertySetEnd();
+
+    return xml;
+  }
+
+  /**
+   * Gets the opening XML Element &lt;defXXXVector&gt; for this Property.
+   *
+   * @return the opening XML Element &lt;defXXXVector&gt; for this Property.
+   */
+  protected abstract String getXMLPropertyDefinitionInit();
+
+  /**
+   * Gets the opening XML Element &lt;defXXXVector&gt; for this Property with a
+   * <code>message</code> to the client.
+   *
+   * @param message A message to be sent to the client.
+   * @return the opening XML Element &lt;defXXXVector&gt; for this Property.
+   */
+  protected abstract String getXMLPropertyDefinitionInit(String message);
+
+  /**
+   * Gets the closing XML Element &lt;/defXXXVector&gt; for this Property.
+   *
+   * @return the closing XML Element &lt;/defXXXVector&gt; for this Property.
+   */
+  protected abstract String getXMLPropertyDefinitionEnd();
+
+  /**
+   * Gets the opening XML Element &lt;setXXXVector&gt; for this Property.
+   *
+   * @return the opening XML Element &lt;setXXXVector&gt; for this Property.
+   */
+  protected abstract String getXMLPropertySetInit();
+
+  /**
+   * Gets the opening XML Element &lt;setXXXVector&gt; for this Property with a
+   * <code>message</code> to the client.
+   *
+   * @param message A message to be sent to the client.
+   * @return the opening XML Element &lt;setXXXVector&gt; for this Property.
+   */
+  protected abstract String getXMLPropertySetInit(String message);
+
+  /**
+   * Gets the closing XML Element &lt;/setXXXVector&gt; for this Property.
+   *
+   * @return the closing XML Element &lt;/setXXXVector&gt; for this Property.
+   */
+  protected abstract String getXMLPropertySetEnd();
+
+  /**
+   * Saves the property and its elements to a file. Ideal to later restore it on
+   * subsecuent executions of the driver.
+   *
+   * @throws IOException
+   */
+  private void saveToFile() throws IOException {
+    File propertiesDir = new File(PROPERTIES_DIR_NAME);
+
+    if (!propertiesDir.exists()) {
+      propertiesDir.mkdir();
+    }
+
+    if (propertiesDir.exists()) {
+      if (propertiesDir.isDirectory()) {
+        File file = new File(propertiesDir, getPropertyNameForFile());
+        FileOutputStream fos = new FileOutputStream(file);
+
+        ObjectOutputStream oos = new ObjectOutputStream(fos);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *oos*. The resource is referenced at lines 524, 526. The resource is closed at line 528. There are other execution paths that do not contain closure statements, for example, when ObjectOutputStream.writeObject throws an exception. Ensure *oos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).